src/Components/Components/src/Services/UriHelperBase.cs

```diff
@@ -70,10 +70,35 @@ public void NavigateTo(string uri, bool forceLoad)
         protected abstract void NavigateToCore(string uri, bool forceLoad);
 
         /// <summary>
-        /// Called to initialize BaseURI and current URI before those values the first time.
-        /// Override this method to dynamically calculate the those values.
+        /// Called to initialize BaseURI and current URI before these values are used for the first time.
+        /// Override this method to dynamically calculate these values.
         /// </summary>
-        protected virtual void InitializeState()
+        public virtual void InitializeState(string uriAbsolute, string baseUriAbsolute)
+        {
+            if (uriAbsolute == null)
+            {
+                throw new ArgumentNullException(nameof(uriAbsolute));
+            }
+
+            if (baseUriAbsolute == null)
+            {
+                throw new ArgumentNullException(nameof(baseUriAbsolute));
+            }
+
+            if (_isInitialized)
+            {
+                throw new InvalidOperationException($"'{typeof(UriHelperBase).Name}' already initialized.");
+            }
+            _isInitialized = true;
+
+            SetAbsoluteUri(uriAbsolute);
+            SetAbsoluteBaseUri(baseUriAbsolute);
+        }
+
+        /// <summary>
+        /// Allows derived classes to lazyly self initialize.
+        /// </summary>
+        protected virtual void Initialize()
```

Comment:
I'm not keen on how we now have both `Initialize` and `InitializeState`. If there truly has to be two different virtual methods, can their names be differentiated more to indicate what's distinctive about their roles? Or could we collapse them both back into one?

Reply:
I called it EnsureInitialized